Update `apply` and add `clear` command

beckyjackson opened the pull request:
Resolves #57 
Resolves #64 

### `apply`

Running `apply` applies the details of one or more [standardized problems tables](#standardized-problems-tables) or [data validation tables](#data-validation-tables) to the spreadsheet as cell formatting and notes.

```
cogs apply [table1.tsv table2.tsv ...]
```

#### Standardized Problems Tables

Standardized problems tables provide a standard table output that can be converted into formatting and notes in the spreadsheet using `apply`. As long as the table follows the format described below, any type of problem can be applied to the sheets. One example is the errors from [ROBOT template](http://robot.obolibrary.org/template).

These tables are applied to the sheets as formats and notes. The three levels of problems will be formatted with a black border and the following backgrounds:
* **error**: light red background
* **warn/warning**: light yellow background
* **info**: light blue background

The notes and formats will be added to any existing, but will take priority over the existing notes and formats.

These tables must have the following headers:
* **ID**: local numeric identifier starting at 1 and counting up
* **table**: name of the table that the problem occurs in
* **cell**: A1 format of problematic cell location
* **level**: severity of the problem; error, warn, or info - this determines the background color of the cell
* **rule ID**: an IRI or CURIE to uniquely identify the problem
* **rule name**: descriptive name of the problem - this is converted to the cell note
* **value**: value of the cell causing problem
* **fix**: can be left blank; a suggested value to replace the problematic value
* **instructions**: can be left blank; detailed instructions on how to fix the problem

#### Data Valildation Tables

The data validation tables are applied to the sheets as data validation rules. These tables must have the following headers:
* **table**: name of the table to add data validation rules to
* **cell**: A1 format of the cell or range of cells to apply data validation rules to
* **condition**: the condition (see below) for the data validation rule
* **value**: the allowed value or values (see bellow)for the data validation rule

For example:

| table  | cell | condition   | value         |
| ------ | ---- | ----------- | ------------- |
| Sheet1 | A2:A | TEXT_EQ     | foo           |
| Sheet1 | B2:B | ONE_OF_LIST | foo, bar, baz |

For full descriptions of each condition type, please see [Google Sheets API ConditionType](https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/other#ConditionType).

Any value greater than one should be specified as a comma-separated list (e.g., for `NUMBER_BETWEEN`, the value could be `1, 10`). For "0" values, leave the `value` column empty.

* *Number Conditions*

| Condition                | Values |
| ------------------------ | ------ |
| `NUMBER_GREATER`         | 1      |
| `NUMBER_GREATER_THAN_EQ` | 1      |
| `NUMBER_LESS`            | 1      |
| `NUMBER_LESS_THAN_EQ`    | 1      |
| `NUMBER_EQ`              | 1      |
| `NUMBER_NOT_EQ`          | 1      |
| `NUMBER_BETWEEN`         | 2      |
| `NUMBER_NOT_BETWEEN`     | 2      |

* *Text Conditions*

| Condition           | Values |
| ------------------- | ------ |
| `TEXT_CONTAINS`     | 1      |
| `TEXT_NOT_CONTAINS` | 1      |
| `TEXT_STARTS_WITH`  | 1      |
| `TEXT_ENDS_WITH`    | 1      |
| `TEXT_EQ`           | 1      |
| `TEXT_IS_EMAIL`     | 0      |
| `TEXT_IS_URL`       | 0      |

* *Date Conditions*

Dates can be supplied in whatever format you like, but we recomment `YYYY-MM-DD`. You can also specify the exact day, a month (`MM-YYYY`), or just a year (`YYYY`). Relative dates (e.g. "today") are not currently supported.

| Condition           | Values |
| ------------------- | ------ |
| `DATE_EQ`           | 1      |
| `DATE_BEFORE`       | 1      |
| `DATE_AFTER`        | 1      |
| `DATE_ON_OR_BEFORE` | 1      |
| `DATE_ON_OR_AFTER`  | 1      |
| `DATE_BETWEEN`      | 2      |
| `DATE_NOT_BETWEEN`  | 2      |
| `DATE_IS_VALID`     | 0      |

* *One-of Conditions*

`ONE_OF_LIST` values should be supplied as a comma separated list. There should be at least two values in the list. For single values, use `TEXT_EQ` instead.

| Condition      | Values |
| -------------- | ------ |
| `ONE_OF_RANGE` | 1      |
| `ONE_OF_LIST`  | 2+     |

* *Other Conditions*

The `CUSTOM_FORMULA` value must be a formula that evaluates to TRUE or FALSE.

| Condition        | Values |
| ---------------- | ------ |
| `BLANK`          | 0      |
| `NOT_BLANK`      | 0      |
| `CUSTOM_FORMULA` | 1      |
| `BOOLEAN`        | 0      |

### `clear`

`clear` removes applied attributes (either from [`apply`](#apply) or manually added to the sheet remotely) from one or more sheets:

```
cogs clear [keyword] [sheet-title-1] [sheet-title-2] ...
```

The keyword must be one of:
* **formats**: sheet formatting
* **notes**: sheet notes
* **rules**: data validation rules
* **all**: formats, notes, and rules

After the keyword, you can supply zero or more sheet titles to remove attributes from. If no sheet titles are provided, the attribute(s) specified by the keyword will be removed from *all* sheets.

Commit:
Add validate command

Commit:
Clean up

Commit:
Fix fetch and push

Commit:
Make validate docs a little clearer

Commit:
Fix clear short form

beckyjackson commented:
This doesn't work well for large sheets because it looks for data validation rules one cell at a time, which causes a "quota exceeded" message.

I'm working on fixing this, but the batch methods in `gspread` aren't well-documented and they aren't returning the data validation information... so I'm marking this as WIP for now.

jamesaoverton commented:
Ok. For IHCC all we really need is to push the data validation rules to Google Sheets, not fetch them.

Commit:
Comment out methods to fetch data validation

beckyjackson commented:
I just commented out the related stuff in `fetch.py`. We can still clear everything using `cogs validate --clear [sheet-title]`.

Commit:
Batch retrieve data validation rules

Commit:
Fix requirements

beckyjackson commented:
I finally figured out how to batch retrieve the data validation (Google API's documentation is... a lot). It should work fine now, but it would be good to get some other people to test this out.

beckyjackson commented:
To do - test out date formatting and see if we can use `YYYY-MM-DD` format instead.

Commit:
Update dates

beckyjackson commented:
@jamesaoverton looks like you can use any format. I've updated the docs to reflect this.

Commit:
Merge branch 'master' into data-validation

Commit:
Fix method names

Commit:
Add validation.tsv to required sheets

beckyjackson commented:
The `validate` command will be merged into `apply`.

Commit:
Merge validate into apply, create clear function

Commit:
Remove validate from test

Commit:
Formatting

beckyjackson commented:
@jamesaoverton 

`apply` has been expanded to accept either standardized problem tables or data validation tables. See the docs for the format of the data validation tables. I changed the headers to be closer to the standardized problems headers so it wouldn't be confusing.

I also added the `clear` function here because I moved the functionality to "clear" formatting from sheets by running `apply` with no args. I don't think it really made sense now that we've made `apply` broader.

I put all the docs in the top comment here.

jamesaoverton commented:
Nice! Please change the keyword "rules" to "validation" -- I think I want to use "rules" for something else.

Commit:
Change rules keyword to validation

beckyjackson commented:
Done!

Commit:
Merge branch 'master' into data-validation

jamesaoverton commented:
In the validation table, the column should be "range" instead of "cell" since it accepts a range.

I'm also thinking that the column names for the the tables that apply accepts should be automatically lowercased before loading, so the user can say "Table" instead of "table".

Commit:
Switch cell to range, make apply headers case-insensitive

Commit:
Update docs

beckyjackson commented:
Makes sense. I've made both these changes.

jamesaoverton commented on cogs/apply.py:
```diff
@@ -1,18 +1,90 @@
+            add_rows.append(rule)
 
+        for row in add_rows:
+            sheet_title = row["Sheet Title"]
```
I don't get this. Is the column named "table" or "Sheet Title"?

beckyjackson replied:
It's "table" in the sheet you pass to "apply" - this is what we decided on for the ROBOT standardized problems tables so I'm following that pattern.

It's "Sheet Title" in the `.cogs/validation.tsv` file because that's the pattern we've used for all the other COGS data files.

It's annoying that it's not the same, but I also think it would be confusing to have inconsistencies between the sets of files.

Commit:
Fix bugs with clearing/applying
